modules/internal/ChapelArray.chpl

```diff
@@ -867,11 +887,33 @@ module ChapelArray {
     return d1._value.dsiEqualDMaps(d2._value);
   }
 
+  // This is temporary while we work on retiring _distribution
+  @chpldoc.nodoc
+  inline operator ==(d1: _distribution(?), d2: record) param {
+    return false;
+  }
+
+  // This is temporary while we work on retiring _distribution
+  @chpldoc.nodoc
+  inline operator ==(d1: record, d2: _distribution(?)) param {
+    return false;
+  }
+
```

Comment:
Basically, if I didn't, existing tests that did comparisons between a Block and Cyclic distribution (say) would just result in an unresolved call error rather than a true/false.  Once all distributions are records and we get rid of the _distribution record these can be removed, and the unresolved comparison is probably the right way to go.  That was just a bigger step than I wanted to take at this point.

Reply:
As an addendum, it completely makes sense that `myBlockDist == myCyclicDist` should lead to a resolution failure, but we have tests that do `myDomain1.dist == myDomain2.dist` and those are the cases I was trying to keep working (for now) with this overload.

Reply:
> it completely makes sense that myBlockDist == myCyclicDist should lead to a resolution failure

I am not bothered too much by the fact that that resolves. Distributions are different types but I can see them being equivalent in some cases. e.g. a Block and a Cyclic array with numLocales elements would be "equal" for some definition of equality. I don't know offhand whether today's `==` return true for a case like that.

Though, after looking at what I typed, the distributions can't be equal because if you have a larger domain, things will differ...

Obviously, not today's discussion, but I can see the type difference here as the type difference between `int` and `real` where an equality between the values might make sense. At least at a high level.

Reply:
The challenge to enabling that vision is: Once we don't have a `dmap`/`_distribution` record, what is the thing that says that `myBlockDist` and `myCyclicDist` should be comparable, but that `myBlockDist` and `myArbitraryRecord` should (presumably) not be?  I.e., what should the where clause be on the `operator ==(r1: record, r2: record) { ... }` to make it only apply to distributions?

Reply:
I don't see anything awkward in the scenario where the distribution types we package with the release have cross-type comparison operators. It may be too much boilerplate with the lack of an overarching record, but design-wise it doesn't feel wrong to me.

This could also be a case for type classes where I can imagine a `distribution` type class that covers all the dists we release. Or maybe a `standardDistribution` type class of sorts. Similarly, there could (actually probably should) be an `interface` where all these types will implement. That can also reduce the boilerplate.

Reply:
I think the maintenance of supporting all-possible distributions is the main downside.  A potential way forward would be to put a where clause on a `==` overload on records that uses `Reflection` to determine whether the record seems to plausibly be a distribution (e.g., supports dsiCreateDomain() or the like).  I'm leaving that for future work, though, once we've got more distributions converted over to this approach.